src/main/anovos/data_analyzer/quality_checker.py

```diff
@@ -721,185 +713,283 @@ def outlier_detection(
+
+    list_odf = []
     for index, i in enumerate(list_of_cols):
-        odf = odf.withColumn(i + "_outliered", F.col("outliered")[index])
-        output_print.append(
-            [
-                i,
-                odf.where(F.col(i + "_outliered") == -1).count(),
-                odf.where(F.col(i + "_outliered") == 1).count(),
-            ]
+        f_composite_outlier = F.pandas_udf(
+            composite_outlier_pandas(params[index]), returnType=T.IntegerType()
+        )
+        odf = odf.withColumn(i + "_outliered", f_composite_outlier(i))
+
+        list_odf.append(
+            odf.select(i + "_outliered")
+            .withColumnRenamed(i + "_outliered", "outliered")
+            .withColumn("attribute", F.lit(str(i)))
         )
 
         if treatment & (treatment_method in ("value_replacement", "null_replacement")):
-            if skewed_cols:
-                warnings.warn(
-                    "Columns dropped from outlier treatment due to highly skewed distribution: "
-                    + (",").join(skewed_cols)
-                )
-            if i not in skewed_cols:
-                replace_vals = {
-                    "value_replacement": [params[index][0], params[index][1]],
-                    "null_replacement": [None, None],
-                }
-                odf = odf.withColumn(
-                    i + "_outliered",
+            replace_vals = {
+                "value_replacement": [params[index][0], params[index][1]],
+                "null_replacement": [None, None],
+            }
+            odf = odf.withColumn(
+                i + "_outliered",
+                F.when(
+                    F.col(i + "_outliered") == 1, replace_vals[treatment_method][1]
+                ).otherwise(
                     F.when(
-                        F.col(i + "_outliered") == 1, replace_vals[treatment_method][1]
-                    ).otherwise(
-                        F.when(
-                            F.col(i + "_outliered") == -1,
-                            replace_vals[treatment_method][0],
-                        ).otherwise(F.col(i))
-                    ),
-                )
-                if output_mode == "replace":
-                    odf = odf.drop(i).withColumnRenamed(i + "_outliered", i)
-            else:
-                odf = odf.drop(i + "_outliered")
+                        F.col(i + "_outliered") == -1,
+                        replace_vals[treatment_method][0],
+                    ).otherwise(F.col(i))
+                ),
+            )
+            if output_mode == "replace":
+                odf = odf.drop(i).withColumnRenamed(i + "_outliered", i)
 
-    odf = odf.drop("outliered")
+    def unionAll(dfs):
+        first, *_ = dfs
+        return first.sql_ctx.createDataFrame(
+            first.sql_ctx._sc.union([df.rdd for df in dfs]), first.schema
+        )
+
+    odf_union = unionAll(list_odf)
+    odf_agg = odf_union.groupby("attribute").pivot("outliered").count()
```

Comment:
Instead of doing it like this, as my experiment with `IV_calculation`, you should run the function(`groupby`, `agg`, etc) inside a loop for every column, then union all the results, instead of union the dataframes first, then run the function. It will help with optimization, as it will reduce the stress on the Spark Session.

An example on what I did on `IV_calculation`:
```
    list_df = []
    for col in list_of_cols:
        df_agg = (
            idf_encoded.select(col, label_col)
            .groupby(col)
            .agg() 
    . . . # All your functions inside
        list_df.append(out_df)
    odf = unionAll(list_df)
```
Reference: [IV_calculation Optimization](https://github.com/anovos/anovos/blob/f2e6cce7d6a5bbd2346b448b788e00402b06a934/src/main/anovos/data_analyzer/association_evaluator.py#L540)
You can test this out to see if there is any improvement.

Reply:
Thanks for pointing out Dat. I have moved the aggregation inside the for loop. The execution time did improve 👍